### PR TITLE
check bucket for meta file even on upload error to increment uploaded…

### LINKED ISSUE #1997: `MimirIngesterHasUnshippedBlocks` can fire even though block was uploaded
#### Describe the bug

`MimirIngesterHasUnshippedBlocks` checks oldest not-yet uploaded block (`cortex_ingester_oldest_unshipped_block_timestamp_seconds`) was generated over 1h ago. But due to a bug in ingester, `cortex_ingester_oldest_unshipped_block_timestamp_seconds` may not be updated even though block was actually uploaded.

In observed case, here is what happened:

- Ingester generated block, and shipper tried to upload it
- Upload of meta.json failed due to 503 error, but meta.json was actually uploaded
- On next shipper iteration, shipper detected that meta.json file already exists, and didn't upload it. In this case shipper considers block to be "uploaded", but it doesn't update `uploaded` counter: https://github.com/grafana/mimir/blob/2c2c65d69bdf466ea66bfa7d4e87704ec9a7f426/pkg/ingester/shipper.go#L157-L164
- Ingester only updates cached version of uploaded blocks if shipper reports any uploaded block. But since it didn't (see previous step), ingester didn't update cached version of shipped blocks: https://github.com/grafana/mimir/blob/0772991cf394d52dff17ccdcb0c52a4652abe706/pkg/ingester/ingester.go#L1797-L1806
- Ingester's `cortex_ingester_oldest_unshipped_block_timestamp_seconds` is based on cached version of shipped blocks.

This is why Mimir may report incorrect `cortex_ingester_oldest_unshipped_block_timestamp_seconds` for another 2 hours, which may cause `MimirIngesterHasUnshippedBlocks` to fire.

#### Expected behavior

No `MimirIngesterHasUnshippedBlocks` if block was actually uploaded.

### PR DESCRIPTION
**Pull request by williamzelesny**
… counter correctly

#### What this PR does

Checks bucket for meta file after getting an upload error incase file was still uploaded in order to increment uploaded counter correctly preventing false alerts

#### Which issue(s) this PR fixes or relates to

Fixes #1997 

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -33,6 +33,7 @@
 * [BUGFIX] Ring: fix bug where instances may appear unhealthy in the hash ring web UI even though they are not. #1933
 * [BUGFIX] API: gzip is now enforced when identity encoding is explicitly rejected. #1864
 * [BUGFIX] Fix panic at startup when Mimir is running in monolithic mode and query sharding is enabled. #2036
+* [BUGFIX] Ingester: check bucket for meta file even on upload error to increment uploaded counter correctly #2059
```
Comment (pstibrany) [inline CHANGELOG.md:36]:
```suggestion
* [BUGFIX] Ingester: check bucket for meta.json file even on upload error to report oldest non-shipped block correctly. #2059
```

**Review by pstibrany — COMMENTED** [review]
Thanks for your PR. Your suggestion would indeed fix the problem when upload returned error, but it actually succeeded.

Since shipper already checks for `meta.json` exists on https://github.com/grafana/mimir/blob/8bff6d023ae1581f61061e4c6fef66cc604aa4c7/pkg/ingester/shipper.go#L148

alternative solution would be to rely on this check (in next call of shipper.Sync, which happens every minute), and just signal to the caller that `meta.Uploaded` has changed (by incrementing `uploaded` or introducing new `metaChanged` return value), so that ingester knows to reload the file.

If you want to proceed with your approach, can you please add a test for your change, fix the log message in this case, and update the comment on line 158?

**Comment by williamzelesny** [comment]
> Thanks for your PR. Your suggestion would indeed fix the problem when upload returned error, but it actually succeeded.
> 
> Since shipper already checks for `meta.json` exists on
> 
> https://github.com/grafana/mimir/blob/8bff6d023ae1581f61061e4c6fef66cc604aa4c7/pkg/ingester/shipper.go#L148
> 
> alternative solution would be to rely on this check (in next call of shipper.Sync, which happens every minute), and just signal to the caller that `meta.Uploaded` has changed (by incrementing `uploaded` or introducing new `metaChanged` return value), so that ingester knows to reload the file.
> 
> If you want to proceed with your approach, can you please add a test for your change, fix the log message in this case, and update the comment on line 158?

Starting to look at writing a test for this but to be honest having trouble wrapping my head around how to go about it.

**Comment by pstibrany** [comment]
> Starting to look at writing a test for this but to be honest having trouble wrapping my head around how to go about it.

For this test to work, you will need to "inject" error to `Upload` call when `meta.json` is being uploaded. You can do that by wrapping existing bucket client, and reimplementing `Upload` method to check for path.

Existing shipper tests use filesystem bucket https://github.com/grafana/mimir/blob/8bff6d023ae1581f61061e4c6fef66cc604aa4c7/pkg/ingester/shipper_test.go#L39. You can keep using it, but add a custom wrapper for it.

Take a look at `failFirstGetBucket` at https://github.com/grafana/mimir/blob/8bff6d023ae1581f61061e4c6fef66cc604aa4c7/pkg/storegateway/bucket_stores_test.go#L715 which does something similar: it fails first `Get` call, and then lets subsequent calls proceed.

**Comment by pracucci** [comment]
@pstibrany Can we consider this PR superseeded by #2435?

**Comment by pstibrany** [comment]
> @pstibrany Can we consider this PR superseeded by #2435?

Yes.

**Pull request closed without merging**